### PR TITLE
[WIP] Checkbox overhaul

### PR DESCRIPTION
**Pull request by eliotwrobson**
Remake of #6785. Overhauls the checkbox element and fixes various issues that have been reported over time.

### CODE REVIEW

File: docs/elements.md [hunk 1]
```diff
@@ -191,6 +191,7 @@ as checkboxes.
 | `min-select`            | integer | special | The minimum number of answers that must be selected in any valid submission. Defaults to `min-correct` if that attribute is specified along with `detailed-help-text="true"`; otherwise, defaults to 1.                              |
 | `max-select`            | integer | special | The maximum number of answers that can be selected in any valid submission. Defaults to `max-correct` if that attribute is specified along with `detailed-help-text="true"`; otherwise, defaults to the number of displayed answers. |
 | `show-number-correct`   | boolean | false   | Display the number of correct choices in the help text.                                                                                                                                                                              |
+| `allow-blank`           | boolean | false   | Whether or not an empty submission is allowed. If `allow-blank` is set to `true`, no response will be marked as incorrect instead of invalid.                                                                                        |
```
Comment (jonatanschroeder) [inline docs/elements.md:194]: This sentence is ambiguous.

```suggestion
| `allow-blank`           | boolean | false   | Whether or not an empty submission is allowed. If `allow-blank` is set to `true`, a response with no selected items will be marked as incorrect instead of invalid.                                                                                        |
```

**Comment by reteps** [comment]
@eliotwrobson Can I work on resolving conflicts for this?

**Comment by eliotwrobson** [comment]
@reteps go ahead! The only caveat is that this is planned to be a large change that will take a significant amount of time to review, so it will likely require another round of conflict resolution later.

**Comment by reteps** [comment]
Backlinks from the previous PR:

- https://github.com/PrairieLearn/PrairieLearn/issues/2590
- https://github.com/PrairieLearn/PrairieLearn/issues/4430
- Add same ordering enum that's being used in the multiple choice refactor. (https://github.com/PrairieLearn/PrairieLearn/issues/5369)
- https://github.com/PrairieLearn/PrairieLearn/issues/793
- Feature request by @mwest1066 on slack:
image
- https://github.com/PrairieLearn/PrairieLearn/issues/878
- https://github.com/PrairieLearn/PrairieLearn/issues/10173

**Comment by eliotwrobson** [comment]
@reteps thanks for reposting! Also, feel free to make changes to this if you have the time. This is slated to be a pretty substantial change to this element, so any work / insights would be much appreciated 🙏

**Comment by reteps** [comment]
If I was recreating this element:

- Add a `type="v2"` attribute
- These attributes would stay the same: `answers-name` `weight` `inline`
- Change `fixed-order` to `order` to mirror other elements.
- Make all booleans positive. `hide-answer-panel -> show-answer-panel`, `hide-letter-keys -> show-letter-keys`, `hide-score-badge -> show-score-badge`, `hide-help-text -> show-help-text`
- Remove `min-correct`, its usage is unclear to me outside of defaults for other elements.
- Rename some properties for clarity `max-correct -> max-display`, `min-select -> min-valid`, `max-select -> max-valid`
- Merge `partial-credit` and partial-credit-method`. I like the term coverage, but `every decision counts` could be `right-wrong` (since every answer is judged independently to be right or wrong. `percent correct` could be `correct-minus-incorrect`, which is more intuitive to me. So the new enum would be `partial-credit="off | coverage | right-wrong | correct-minus-incorrect"` (with abbreviations)
- Merge the help text options `detailed-help-text` and `show-number-correct`. `help-text-mode="num-valid | num-correct | all"`

**Comment by reteps** [comment]
@jonatanschroeder The v="2" would be more helpful for documentation than actual implementation. With so many different fields, it would be nice to present a tabbed interface "here is the old version of the element and here is the new version, and this is how to upgrade". The v="2" would be an optional field that throws an error if you use any deprecated fields.

**Comment by nwalters512** [comment]
> The v="2" would be more helpful for documentation than actual implementation. With so many different fields, it would be nice to present a tabbed interface "here is the old version of the element and here is the new version, and this is how to upgrade".

Why does `v="2"` have to exist for that? Wouldn't a table of "these attributes are deprecated, here's what they did, here's what to use instead" work just fine?

> The v="2" would be an optional field that throws an error if you use any deprecated fields.

As a user, what's my incentive for _ever_ setting this attribute? This would just make a bunch of work for me and wouldn't actually change anything about how the element behaves, right? If I'm using a library or an application, my incentive to upgrade through breaking changes is in theory new features and bug fixes, but presumably we're not going to actually maintain two separate versions of this element so I don't think that applies here?

More questions: how does a potential `version` attribute fit into our wider element development strategy? When does one decide to cut a new "version" of something, vs. just deprecate/rename-with-backwards-compat/etc. like we already do? When changing one attribute? Many attributes? When it feels right? Does having versioned elements make things more complicated for LLMs?

**Comment by reteps** [comment]
Thanks for the feedback @nwalters512 @jonatanschroeder . I see how the `v="2"` property wouldn't be useful. In this case, we can map the old properties to new properties.

I do think that versioning is still an important piece of writing APIs for these components. As long as there is a clear way to document these API changes and it can be made backwards-compatible in a straightforward way, I see why we wouldn't want versioning.

> More questions: how does a potential version attribute fit into our wider element development strategy? When does one decide to cut a new "version" of something, vs. just deprecate/rename-with-backwards-compat/etc. like we already do? When changing one attribute? Many attributes? When it feels right? Does having versioned elements make things more complicated for LLMs?

I'm interpreting this as a set of thought points. Here is why I wanted to bring up versioning:

(1) we historically haven't upgraded or changed elements much AFAIK
(2) instructors don't change or upgrade elements much, they are too busy with other things
(3) because of this, the "wider element development strategy" is 'maintain backwards compatibility'. If you have more resources on what exactly the current element development strategy is, I would be interested to see it.
(4) The elements are going to change more quickly, and add more features, now that PrairieLearn can hire additional developers. There are many feature requests for element changes already.
(5) At some point, this is going to cause an backwards-incompatible change.
(6) Because of this, we should start thinking about strategies for element upgrades, deprecations, versioning, etc. before it becomes an actual limitation and starts slowing down the pace at which we can improve things. 


Here is a pretty weak argument for a potential backwards-incompatible change:

- I just suggested via slack to someone that if elements provide classes that they can style, that they can modify the styling with external CSS. Let's say I had done this for `pl-checkbox`.
- Now, lets say I implement some UI changes to the element to support "None of the above" as in https://ux.stackexchange.com/a/67658. This causes me to refactor the element in such a way that the original classes changed because the HTML layout is fundamentally different.
- In this case, what if the user wants the legacy version of the element that works with their styling, rather than the new version that doesn't? Are they just out of luck?

tldr - I would love to have a meeting and discuss what the element development strategy is, preferably over a call 😄

**Comment by eliotwrobson** [comment]
A little bit tangential to the point of this PR, but to the points you outlined @reteps:
- (3) is always going to be a priority (not just for elements but in general) since we don't want to break potentially large question banks that instructors don't have the resources to actively maintain amidst many changes.
- I don't think (4) is a valid assumption going forward, and I actually think the emphasis is going to move away from element changes in the near to medium term, since there are proposed features for other parts of the platform that will have more impact. In either case, element upgrades and versioning don't apply to completely new elements, which is most of our feature requests right now.

**Comment by nwalters512** [comment]
> The elements are going to change more quickly, and add more features, now that PrairieLearn can hire additional developers. There are many feature requests for element changes already.

We'll definitely add more features; adding features doesn't require versioning. But I'll re-emphasize what Eliot said: many of the things people want (or that we want to build) are actually brand-new elements, not iterations on existing ones.

> At some point, this is going to cause an backwards-incompatible change.

No. Our changes by necessity must be backwards compatible. One of the most important responsibilities we have to our users is that we _do not break existing questions_. Even with a proposed versioning scheme, we're not allowed to make an actual backwards-incompatible change. We can introduce _new_ functionality, but we cannot break anything already in use by questions. Ever. (I'll add the caveat that I'm talking only about the element API, not implementation details like DOM structure - more on that in a second.)

If we're actually making a backwards-incompatible, it means we've proved to ourselves that no question is using the thing we're changing, at that that point there's no reason to introduce a version, or treat it carefully in the documentation, or whatever.

> - I just suggested via slack to someone that if elements provide classes that they can style, that they can modify the styling with external CSS. Let's say I had done this for pl-checkbox.
> - Now, lets say I implement some UI changes to the element to support "None of the above" as in https://ux.stackexchange.com/a/67658. This causes me to refactor the element in such a way that the original classes changed because the HTML layout is fundamentally different.

You just discovered precisely why we don't have `class` attributes today and why we're not likely to ever add them: we can't introduce that without effectively making the shape of our DOM part of our public API, which we are not interesting it doing exactly because it makes it hard for us to change things.

We can discuss this on a call. I may be convinced otherwise, but IMO we're spending a lot of time discussing something that provides no immediate benefit, and little-to-no benefit in the long term, and that does not solve any problems we have faced before or are facing now.

**Comment by reteps** [comment]
I think I didn't communicate what I meant by backwards-incompatible change: I meant that if we need to introduce one, we would add a new version. I never want to break anyone's question.

> You just discovered precisely why we don't have class attributes today and why we're not likely to ever add them: we can't introduce that without effectively making the shape of our DOM part of our public API, which we are not interesting it doing exactly because it makes it hard for us to change things.

Isn't this the exact thing that a `v="2"` field would fix? Then you can render the old template, or the new template depending on this? I completely agree with the fact that there should never be a backwards-incompatible change.

> we should start thinking about strategies for element upgrades, deprecations, versioning, etc. before it becomes an actual limitation and starts slowing down the pace at which we can improve things.

Isn't this exactly what you are saying? We don't want to add a feature because it would slow down the speed that we can iterate on the element? Granted this particular feature is a weak one and I wouldn't want to actually add it, or think that it's a priority currently _to_ add it.

**Comment by nwalters512** [comment]
Gotcha; yeah, I think we're thinking about this a bit differently, but I'm glad we're on the same page regarding "don't break things"!

> Isn't this the exact thing that a v="2" field would fix? Then you can render the old template, or the new template depending on this? I completely agree with the fact that there should never be a backwards-incompatible change.
> ...
> Isn't this exactly what you are saying? We don't want to add a feature because it would slow down the speed that we can iterate on the element? Granted this particular feature is a weak one and I wouldn't want to actually add it, or think that it's a priority currently to add it.

For `class` specifically, even if versioning were in play, that feature would slow us down _forever_. It's not a one-time version bump. Now, anytime we make _any_ change to DOM structure, we have to consider if it could break users, and if so, introduce a new version, figure out how to document what it means to users, maintain an additional copy of the template, etc.

All this versioning would actually make it harder for end users to use the element. Say someone is using an unversioned version of the element, then we change the DOM structure three times (and bump the version three times), then we introduce a `foobar` attribute. If a user wants to use `foobar`, now they have to parse through what it means to jump forwards three versions (even though they're probably not impacted). Or we have to backport `foobar` to all versions, which in its own right would slow us down a bunch.

If we maintain strict backwards-compatibility and just add/evolve existing features, using `foobar` (or any new bit of functionality) is basically trivial: you just use it, and things just work.

Do you have other examples of potential backwards-incompatible changes you think we'd want to make that would benefit from versioning? I find it more helpful to talk about specifics; the `class` example was really useful 🙏

**Comment by reteps** [comment]
More related to the actual PR, can we split the changes up from https://github.com/PrairieLearn/PrairieLearn/pull/10524#issuecomment-2840906711 into smaller PRs that are easier to review?

**Comment by nwalters512** [comment]
Yes please; I'm always in favor of smaller PRs when it's feasible 🙏

**Comment by reteps** [comment]
> Do you have other examples of potential backwards-incompatible changes you think we'd want to make that would benefit from versioning? I find it more helpful to talk about specifics; the class example was really useful 🙏

I think that this is worth revisiting at a future point in time, as I just wanted to discuss it before it becomes an issue; which it isn't yet. 

>  Say someone is using an unversioned version of the element, then we change the DOM structure three times (and bump the version three times), then we introduce a foobar attribute.

Yes, this particular feature isn't a great example exactly because of this issue.

Another feature I was thinking of was element extensions for custom grading strategies, though that isn't necessarily backwards-incompatible either.

**Comment by eliotwrobson** [comment]
> More related to the actual PR, can we split the changes up from [#10524 (comment)](https://github.com/PrairieLearn/PrairieLearn/pull/10524#issuecomment-2840906711) into smaller PRs that are easier to review?

If this can be done in a way where the intermediate changes can be independently deployed. The problem with a lot of the element work is that we rely mostly on manual testing to verify behavior (this maybe isn't the best strategy but it is what it is for now). Having to do this to verify intermediate versions of the element is more work than just doing it once all of the planned changes have been implemented. This is a case where a stacked PR or similar might be a good strategy.

**Comment by mwest1066** [comment]
Another strategy we could consider at some point when making changes to elements (and JSON formats) is to have something like https://codemod.com that offers users a 1-click upgrade for their code. I'm thinking the UI could be something like:
* Instructors get a banner saying "we can upgrade some of the code in your course", with buttons to "Just do it" or "See details".
* We'd only do this if we are certain that the new code has exactly the same functionality as the old. I'm thinking of things like changing defaults on attributes, changing the names of things, etc.
* Any time an instructor uses a GUI or AI to edit JSON or code we could automatically upgrade things for them without asking.
* At some point we could auto-apply such changes, e.g., if the instructor has not clicked the "upgrade" button in N months.

The main benefits of this are:
* For users, it would make their code look like the current docs. At the moment, we don't document deprecated attributes, which can make it confusing when you look at an old bit of code and try to understand it. This would also help our AI.
* For us, we can simplify some of our code once we are sure all course code has been upgraded. But see below for a caution about this.

The main downsides I see of this strategy are:
* Users bringing course code over from an old self-hosted PL might find that it's too out of date to work. At the moment this doesn't happen because we have strict backwards compatibility. To avoid this issue we could continue to support the old formats indefinitely, while still allowing users to upgrade to improve their experience.
* Some users might not like the idea that they need to change their code to keep it up to date. I'm probably too steeped in the NodeJS ecosystem where regular upgrades are just part of life, but to me it feels reasonable that you need change to stay current.

Note that I think this is a very low priority at present. I currently think our "just maintain backwards compatibility" strategy is working well. I'm merely raising this idea here as another aspect to consider if we find backwards compatibility is starting to become a drag on making improvements.

**Comment by reteps** [comment]
> If I was recreating this element:
> 
> * Add a `v="2"` attribute

Ignore

> * These attributes would stay the same: `answers-name` `weight` `inline`

Ignore

> * Change `fixed-order` to `order` to mirror other elements.

Done

> * Make all booleans positive. `hide-answer-panel -> show-answer-panel`, `hide-letter-keys -> show-letter-keys`, `hide-score-badge -> show-score-badge`, `hide-help-text -> show-help-text`

Too big of a backwards-compat issue, ignoring.

> * Remove `min-correct`, its usage is unclear to me outside of defaults for other elements.

Ignored, I misunderstood.

> * Rename some properties for clarity `max-correct -> max-display`, `min-select -> min-valid`, `max-select -> max-valid`

Too big of a backwards-compat issue, ignoring.

> * Merge `partial-credit` and `partial-credit-method`. I like the term coverage, but `every decision counts` could be `right-wrong` (since every answer is judged independently to be right or wrong). `percent correct` could be `correct-minus-incorrect`, which is more intuitive to me. So the new enum would be `partial-credit="off | coverage | right-wrong | correct-minus-incorrect"` (with abbreviations). Maybe this can be called `grading-strategy` instead?

This PR should do that. It will map `false` to `off` and `true` to `correct-minus-incorrect` as suggested by @jonatanschroeder .

> * Merge the help text options `detailed-help-text` and `show-number-correct`. `show-help="num-valid | num-correct | all"`

I think this could be addressed in a follow up PR if we want.

> * Add a feedback mode suggested in [Adding Feedback for pl-checkbox problems #10173](https://github.com/PrairieLearn/PrairieLearn/issues/10173). `show-feedback="selected | correct | incorrect | unselected"`.

This PR shouldn't do that, but this feels useful.

> * Implement "None of the above" like this: https://ux.stackexchange.com/a/67658
> 

This PR shouldn't do that, but this feels useful.

> Additional features to consider:
> 
> * Add a `weight` to `pl-answer` -- some checkboxes are more important to get right than others.

This PR shouldn't do that, but this feels useful.

File: apps/prairielearn/elements/pl-checkbox/pl-checkbox.py [hunk 2]
```diff
@@ -169,48 +169,86 @@ def generate_insert_text(
...
-    )
+    if not pl.has_attrib(element, "partial-credit"):
+        # No attribute specified, use default
+        return PARTIAL_CREDIT_DEFAULT
+
+    # Try to parse as enum first (new style)
+    partial_credit_str = pl.get_string_attrib(element, "partial-credit", "")
+
+    # Handle backward compatibility: old boolean + method style
+    if partial_credit_str.lower() in ["true", "false"]:
```
Comment (reteps) [inline apps/prairielearn/elements/pl-checkbox/pl-checkbox.py:205]: This is too restrictive. `get_boolean_attribute` allows more values.

**Comment by reteps** [comment]
Hmm, I didn't mean to do that...

**Comment by reteps** [comment]
Closing in favor of https://github.com/PrairieLearn/PrairieLearn/pull/10524. @eliotwrobson wanted a cleaner commit history.
